project/utils/include/utils/cuda.cuh

```diff
@@ -0,0 +1,247 @@
+         * @param[in] line  The line that caused the error.
+         * 
+         * @return the generated message.
+         */
+        static std::string message(cudaError_t error, const char file[], int line)
+        {
+            std::stringstream string;
+            string << "in " << file << " on line #" << line << ": "
+                   << "(" << error << "): " << cudaGetErrorString(error);
+            return string.str();
+        }
+    };
+
+    /**
+     * Call a CUDA function and handle any errors by throwing a CudaException.
+     */
+    #define CUDA_CALL(expression)                                  \
+    do                                                             \
+    {                                                              \
+        cudaError_t error = (expression);                          \
+        if (error != cudaSuccess)                                  \
+        {                                                          \
+            throw utils::CudaException(error, __FILE__, __LINE__); \
+        }                                                          \
+    } while (false)
+
+    //--------------------------------------------------------------------------
+    // Kernels
+    //--------------------------------------------------------------------------
+
+    /**
+     * Describes a function that launches a kernel on a range of data.
+     * 
+     * @param[in] offset      The offset into the data that should be passed to kernel.
+     * @param[in] grid_size   The grid size to use in the kernel call.
+     * @param[in] block_size  The block size to use in the kernel call.
+     */
+    using KernelLaunchFunction = std::function<void(size_t offset, 
+                                                    unsigned grid_size, 
+                                                    unsigned block_size)>;
+
+    /**
+     * Launch a kernel using the most efficient grid sizes and block sizes as
+     * reported by the hardware.
+     * 
+     * @tparam KernelFunction The type of the kernel function.
+     * 
+     * @param[in] kernel         The kernel to be launched.
+     * @param[in] threads_needed The number of threads that need to be launched.
+     * @param[in] launch_kernel  A function to launch the kernel.
+     * 
+     * @c threads_needed does not need to be rounded; it can just be the size of
+     * the data. This function will perform any appropriate rounding/padding.
+     * However, note that either @c launch_kernel or @c kernel will need to
+     * perform checks to make sure that @c kernel does not perform an 
+     * out-of-bounds memory access in case more threads are launched than are
+     * needed.
+     * 
+     * @throws CudaException if a CUDA error occurs.
+     */
+    template<typename KernelFunction>
+    void launch_kernel(KernelFunction kernel, 
+                       size_t threads_needed, 
+                       const KernelLaunchFunction& launch_kernel)
```

Comment:
I didn't even realize I did this lol.

Taking another look at this function, though, I think I can eliminate the launch function altogether. I originally had it because I thought I might need to split the kernel call into multiple calls to handle large amounts of data, but that obviously didn't seem to be the case (at least, for the data sizes I tested).

Reply:
Yeah I kind of figured part of this architecture was maybe just leftover - similarly, it looks like the `offset` parameter in the `KernelLaunchFunction` typedef only ever gets the value 0.